sarif/filter/general_filter.py

```diff
@@ -0,0 +1,207 @@
+        self.exclude_filters = exclude_filters
+        self.apply_exclusion_filter = len(exclude_filters) > 0
+
+    def rehydrate_filter_stats(self, dehydrated_filter_stats, filter_datetime):
+        """
+        Restore filter stats from the SARIF file directly,
+        where they were recorded when the filter was previously run.
+
+        Note that if init_filter is called,
+        these rehydrated stats are discarded.
+        """
+        self.filter_stats = load_filter_stats_from_json_camel_case(
+            dehydrated_filter_stats)
+        self.filter_stats.filter_datetime = filter_datetime
+
+    def _zero_counts(self):
+        if self.filter_stats:
+            self.filter_stats.reset_counters()
+
+    def _filter_append(self, filtered_results: List[dict], result: dict):
+        # Remove any existing filter log on the result
+        result.setdefault("properties", {}).pop("filtered", None)
+
+        matched_include_filters = []
+        if self.apply_inclusion_filter:
+            matched_include_filters = \
+                self._filter_result(result, self.include_filters)
+            if not matched_include_filters:
+                return
+
+        if self.apply_exclusion_filter:
+            if self._filter_result(result, self.exclude_filters):
+                self.filter_stats.filtered_out_result_count += 1
+                return
+
+        included = {
+            "state": "included",
+            "matchedFilter": matched_include_filters,
+        }
+        self.filter_stats.filtered_in_result_count += 1
+        included["filter"] = self.filter_stats.filter_description
+        result["properties"]["filtered"] = included
+
+        filtered_results.append(result)
+
+    def _filter_result(self, result: dict, filters: List[str]) -> List[dict]:
+        matched_filters = []
+        if filters:
+            # filters contains rules which treated as OR.
+            # if any rule matches, the record is selected.
+            for filter_spec in filters:
+                # filter_spec contains rules which treated as AND.
+                # all rules must match to select the record.
+                matched = True
+                for (prop_path, prop_value_spec) in filter_spec.items():
+                    resolved_prop_path = \
+                        FILTER_SHORTCUTS.get(prop_path, prop_path)
+                    jsonpath_expr = jsonpath_ng.ext.parse(resolved_prop_path)
+
+                    found_results = jsonpath_expr.find(result)
+                    if found_results:
+                        value = found_results[0].value
+                        value_spec = \
+                            resolve_field_shortcuts(resolved_prop_path,
+                                                    prop_value_spec)
+                        filter_function = get_filter_function(value_spec)
+                        if filter_function(value):
+                            continue
+                    matched = False
```

Comment:
I've retested the old functionality for comparison.

Running the blame filter on a file with some blame info but not complete, it does include the records that lack blame filter info:

> Results were filtered by 'Sam only' at Sat Sep 30 07:50:09 2023: 885 filtered out, 35 passed the filter, 2 included by default for lacking blame data to filter

If there is no blame info at all, this applies to all the records:

> Results were filtered by 'Sam only' at Sat Sep 30 07:52:57 2023: 0 filtered out, 0 passed the filter, 920 included by default for lacking blame data to filter

This was a good choice for blame info, but might not be good for other filters.  I think it would be good to have a way to specify `strict` or not.  I think a top level `strict: True` element in the filter YAML (defaulting to false if absent) is the easiest way.  More complex alternatives are to be able to specify `strict` within each `include` block, or a new `--filter-strict` command line flag.

Reply:
Yeah, makes sense. Let's add something a global configuration and ability to override it in every rule, like this:
```yaml
description: Test filter
configuration:
  strict: true
include:
  - author-email:
      value: "@microsoft.com"
      strict: false
```

Reply:
I think I prefer `default-include: true` - i.e. the default action is to include if you can't filter.

Reply:
I implemented `default-include` configuration option. Added a counter back for default included if missing data.
Renamed mentioning of 'field' to 'property' to match SARIF spec.

Reply:
Great, ready to merge now :-)